File: Typescript Solutions/RansomNote.ts
interface IHash {
    [key: string]: number;
}

function canConstruct (ransomNote: string, magazine: string) : boolean {
    let mag_dict: IHash = {};

    for(let i = 0; i < magazine.length; i++) {
        if (mag_dict[magazine[i]] == undefined) {
            mag_dict[magazine[i]] = 0;
        }

        mag_dict[magazine[i]] += 1;
    }

    for(let i = 0; i < ransomNote.length; i++) {
        if (mag_dict[ransomNote[i]] == undefined) {
            mag_dict[ransomNote[i]] = 0;
        }

        mag_dict[ransomNote[i]] -= 1;

        if (mag_dict[ransomNote[i]] < 0) {
            return false;
        }
    }

    return true;
}

File: C# Solutions/RansomNote.cs
using System;
using System.Collections.Generic;

namespace RansomNote
{
    class Program
    {
        public static Program Instance = new Program();
        
        static void Main(string[] args)
        {
            Console.WriteLine(Instance.canConstructNote("aa", "aab"));
        }

        public bool canConstructNote(string note, string magazine)
        {
            Dictionary<char, int> mag_dict = new Dictionary<char, int>();
            int value;

            foreach (char letter in magazine)
            {
                if (mag_dict.TryGetValue(letter, out value))
                {
                    mag_dict[letter] = value + 1;
                }
                else
                {
                    mag_dict.Add(letter, 1);
                }
            }

            foreach (char letter in note)
            {
                if(mag_dict.TryGetValue(letter, out value))
                {
                    mag_dict[letter] = value - 1;

                    if (mag_dict[letter] < 0)
                    {
                        return false;
                    }
                }
                else
                {
                    return false;
                }
            }

            return true;
        }
    }
}